### ISSUE TITLE
X.2 measured: NVScoreVariants cannot be bit-identical, and CNNScoreVariants no longer exists

### ISSUE THREAD

**Issue by BenjaminDEMAILLE** [IPNP-BIPN/gatk-rs#250]
Measurements taken on this laptop against GATK 4.6.2.0's own model file, before any port is written. Three findings, each of which changes what #73 asks for.

## 1. CNNScoreVariants is gone, and TensorFlow with it

`DeprecatedToolsRegistry` retires it as of **4.6.1.0**:

> `deprecatedTools.put("CNNScoreVariants", Pair.of("4.6.1.0", "Please use the replacement tool NVScoreVariants instead, which produces virtually identical results"));`

`CNNVariantTrain` is retired too. `scripts/gatkcondaenv.yml.template` contains **no TensorFlow and no Keras**: the ML surface of the reference is one tool, `NVScoreVariants`, on PyTorch 2.1.0 with an MKL BLAS. The title of #73 names a tool that the version we reproduce does not ship.

## 2. The reference does not reproduce itself, so "bit-identical" is unreachable here

The same shipped model (`small_2d.pt`), the same deterministic input, `eval()` so dropout is off:

| comparison | floats differing | worst gap |
|---|---|---|
| CPU torch 2.1.0 vs CPU torch 2.10.0 | 6 of 8 | **1 ULP** |
| CPU vs MPS, torch 2.1.0 | 7 of 8 | 4 ULP |
| CPU vs MPS, torch 2.10.0 | 7 of 8 | 4 ULP |
| CPU rerun, same version | **0 of 8** | 0 |

Each run is deterministic; the differences are the kernels, not noise. Ten minor versions of PyTorch move the last bit of six outputs out of eight, and the Apple GPU moves four bits.

So the question this milestone poses, "reproduce the exact kernels and accumulation order", has no answer that survives a PyTorch upgrade: **the reference disagrees with itself at 1 ULP across its own versions**. What is left to decide is the tolerance, and these numbers give its scale: about 1e-06 on logits of order 1.

That also flattens the library comparison. `tch` was the only candidate that could share GATK's kernels; once 1 ULP is accepted, that advantage disappears, and `tract` (pure Rust, no native runtime) or `ort` (ONNX Runtime, bundled) become comparable candidates to be measured the same way.

## 3. The shipped artefacts are pickled Python modules

`1d_cnn_mix_train_full_bn.pt` (6.4 MB) and `small_2d.pt` (2.0 MB) are zip archives holding `archive/data.pkl` and raw tensors, with **no `code/` directory**: they are neither TorchScript nor state_dicts. `torch.load` returns a live `GATK_CNN_2D` object, which unpickles only because the class is importable. The artefact references Python code, not just numbers.

Consequences for every route considered:

| route | what the pickle forces |
|---|---|
| `tch` + rebuild in `tch::nn` | extract tensors from the pickle, then transcribe the architecture |
| `tch` + `CModule::load` | **impossible as shipped**: there is no TorchScript in the file |
| `ort` / `tract` | an offline ONNX conversion, which is a Python build step |
| pure-Rust frameworks | same extraction, and kernels further from the reference |

No route reads the file as it stands, including staying inside PyTorch.

## What this milestone should now ask

1. Set a tolerance for `NVScoreVariants` and justify it from these numbers, or quarantine the tool and report it as bio-identical, which the milestone text already allows for.
2. Pick the runtime on grounds other than bit-identity, since none is available: dependency weight, licence, and whether an ONNX conversion step is acceptable.
3. Correct the scope: one tool, not two.

The probe is `scratchpad/mps_probe.py`, not committed; it can be turned into a suite if the tolerance route is taken.

Parent: #73.

**Comment by BenjaminDEMAILLE** [comment]
Checked the obvious hole in the measurement above: the inputs come from `torch.rand` with a fixed seed, so if the generator had changed between versions the comparison would have been meaningless.

**It has not.** The input tensors are bit-identical across 2.1.0 and 2.10.0: 0 of 5014 floats differ. The kernel conclusion stands.

The check produced a shorter proof of the same point, though. Two control lines from the same run:

```
2.1.0    reads.sum() = 245601.625        annotations.sum() = 6.512181758880615
2.10.0   reads.sum() = 245601.609375     annotations.sum() = 6.512181758880615
```

Identical bytes in, different sum out. No convolution, no model, no GPU: a plain `sum()` over 491,520 floats, where the reduction tree changed between versions and the answer moves by two last bits. The 14-element sum beside it agrees exactly, because it fits in one block and needs no tree.

That is the whole mechanism in one line: floating-point addition is not associative, the grouping is an implementation detail, and PyTorch revises it freely and documents no bitwise guarantee across versions, devices or thread counts.

It also widens the scope of the finding. The divergence is not specific to this model or to Metal; it is in the reduction itself, so it would appear in any framework that blocks its sums differently, which is every one of them.

**Comment by BenjaminDEMAILLE** [comment]
**Direction taken: latest libtorch, not GATK's pinned 2.1.0.**

Checked what is actually distributed, since the earlier note repeated a stale error message from `torch-sys` ("Pre-built version of libtorch for apple silicon are not available"), which has not been true since 2.2.0:

| archive | macOS arm64 | Linux x86-64 |
|---|---|---|
| 2.1.0 (GATK's pin) | **404** | available |
| 2.2.0 | available, 55 MB | available |
| 2.11.0 (what `tch` 0.24 requires) | available, 74 MB | available |
| 2.13.0 (latest) | available | available |

So `tch` 0.24 with `download-libtorch` gives **zero Python on every platform we use**, including Apple Silicon: no conda, no pip, no wheel extraction, one 74 MB archive fetched at build time. The only version with no macOS archive at all is 2.1.0, which is precisely the one GATK pins — a reason to move forward rather than to pin backwards.

The cost is already measured and accepted: about 1 ULP against GATK's own 2.1.0 on CPU, on an output whose bit-identity is unreachable anyway, since two PyTorch versions disagree with each other by that same margin.

This settles the runtime question for `NVScoreVariants`. What remains open in this issue is the tolerance to declare, and whether the ONNX route (`tract`, no native runtime at all) is worth measuring beside it.

**Comment by BenjaminDEMAILLE** [comment]
**Runtime chosen: `tch` (libtorch bindings). The ONNX routes are measured and set aside.**

Both alternatives were measured on the shipped model rather than argued about, exporting it to ONNX at opset 17, which succeeded with no unsupported operator:

| execution | floats differing | worst gap | absolute |
|---|---|---|---|
| PyTorch CPU | reference | — | — |
| ONNX Runtime, CPU provider | 7 of 8 | 3 ULP | 1.9e-06 |
| PyTorch MPS | 7 of 8 | 4 ULP | 1.4e-06 |
| ONNX Runtime, CoreML provider | 8 of 8 | **7131 ULP** | **3.1e-03** |

Two things worth keeping from that.

ONNX Runtime on CPU lands at 3 ULP, inside the same envelope as everything else, which is a useful independent confirmation that ~1e-06 is the right scale for a tolerance whatever the engine. It was not rejected for accuracy.

CoreML is three orders of magnitude out, at 3e-03. That is not reassociation, it is a change of precision: CoreML runs parts of the graph in reduced precision, and the session log shows the graph split into two partitions with 18 of 20 nodes taken. Unusable for conformance even under tolerance. Note that the Apple hardware therefore behaves very differently depending on the path: **MPS through PyTorch is 4 ULP and acceptable; CoreML through ORT is 3e-03 and is not**.

**Decision: stay on `tch`.** It keeps the model in its native form, needs no offline conversion step and no generated artefact to justify, and its accelerator path on this machine is the one that stays within tolerance.

Constraint that comes with it: `tch` is at 0.24.0 (March 2026) and requires **libtorch 2.11.0**; libtorch archives exist up to 2.13.0, so 2.13 waits on a `tch` release rather than on us. 2.11 gives zero Python on every platform we use, Apple Silicon included.

Op coverage was checked on both shipped models, not assumed: Conv1d, Conv2d, ReLU, Dropout, MaxPool2d, BatchNorm1d, Linear, cat, transpose, permute, flatten. All run on MPS with no fallback, at 3 to 4 ULP.

**Comment by BenjaminDEMAILLE** [comment]
## libtorch access, measured on 2026-08-15

Three things this changes, all measured on this laptop against torch **2.10.0** (latest available; the reference's `pytorch=2.1.0` pin is not inherited).

### 1. The `CModule::load` row of the table above is wrong

`tch` + `CModule::load` is recorded as "**impossible as shipped**: there is no TorchScript in the file". True of the file as shipped, but both artefacts trace, and the trace is verifiable:

| artefact | `torch.jit.trace` | reload reproduces the logits |
|---|---|---|
| `small_2d.pt` | ok | **bit-identical** |
| `1d_cnn_mix_train_full_bn.pt` | ok | **bit-identical** |

The traced archives carry a real `code/` directory (bytecode version 3). So the route is a one-off conversion whose fidelity is checkable, not a transcription of the architecture. `torch.onnx.export` fails here for want of `onnxscript`, so the ONNX routes cost one more Python dependency than the TorchScript route does.

### 2. The operation surface is eleven ordinary ops

Enumerated from the traced graphs rather than read off the architecture:

- `small_2d`: `_convolution` ×4, `relu` ×6, `max_pool2d` ×2, `batch_norm`, `linear` ×3, `cat`, `flatten`, `permute` ×2, `dropout`/`feature_dropout` ×5
- `1d`: `_convolution` ×3, `relu` ×5, `batch_norm`, `linear` ×3, `cat` ×2, `flatten`, `transpose`, `dropout`

All core. Dropout is identity under `eval()`. **Nothing in these models steers the runtime choice**, which removes the last capability-based argument from the comparison.

### 3. `tch` does not build on macOS arm64, and the reason is not what I first thought

| attempt | result |
|---|---|
| `tch` 0.17 + pip wheel (`LIBTORCH_USE_PYTORCH=1`) | build script cannot find torch from its env |
| `tch` 0.17 + pip wheel via `LIBTORCH=` | ATen signature mismatches: 0.17 targets an older libtorch |
| `tch` 0.24 (latest) + pip wheel | `no member named 'miopen_ctc_loss' in namespace 'torch'` |
| `tch` 0.24 + **official** libtorch 2.10.0 macOS arm64 (73 MB, downloaded) | **same failure** |

My hypothesis was that the pip wheel ships a header subset and the official distribution would fix it. **The download disproved it**: the official macOS arm64 distribution declares `miopen_ctc_loss` exactly as often as the wheel does — zero times. `torch-sys` generates bindings for every aten op including the ROCm/MIOpen ones, and no macOS libtorch declares them, ROCm being Linux-only.

Consequence: **`tch` is a Linux-side dependency**. That is survivable, because the bit-identity claim is CI-gated and CI is Linux x86-64, but it means a `tch` route cannot be developed or debugged locally on this machine without a container. `ort` and `tract` do not have this problem and are worth re-measuring on that ground alone — which is a packaging argument, not the capability argument this issue already dismissed.

### What is now decided and what is not

- **Decided**: latest versions throughout; the tolerance question stands as this issue framed it; op support is not a discriminator.
- **Open**: whether a Linux-only native dependency is acceptable for one tool, against a pure-Rust or bundled-runtime alternative that costs an offline ONNX conversion step.

The probes are in the session scratchpad and are not committed.
